compiler/src/dotty/tools/dotc/transform/CompleteJavaEnums.scala

```diff
@@ -114,37 +113,42 @@ class CompleteJavaEnums extends MiniPhase with InfoTransformer { thisPhase =>
 
   private def isJavaEnumValueImpl(cls: Symbol)(using Context): Boolean =
     cls.isAnonymousClass
-    && ((cls.owner.name eq nme.DOLLAR_NEW) || cls.owner.isAllOf(EnumCase))
-    && {
-      val enumCls = cls.owner.owner.linkedClass
-      enumCls.derivesFromJavaEnum && enumCls.is(Enum)
-    }
+    && (((cls.owner.name eq nme.DOLLAR_NEW) && cls.owner.isAllOf(Private|Synthetic)) || cls.owner.isAllOf(EnumCase))
+    && cls.owner.owner.linkedClass.derivesFromJavaEnum
+
+  @threadUnsafe
+  private lazy val enumCaseOrdinals: mutable.Map[Symbol, Int] = mutable.AnyRefMap.empty
+
+  private def registerEnumClass(cls: Symbol)(using Context): Unit =
+    cls.children.zipWithIndex.foreach(enumCaseOrdinals.put)
+
+  private def ordinalFor(enumCase: Symbol): Int = enumCaseOrdinals.remove(enumCase).get
 
   /** 1. If this is an enum class, add $name and $ordinal parameters to its
    *     parameter accessors and pass them on to the java.lang.Enum constructor.
    *
-   *  2. If this is an anonymous class that implement a value enum case,
+   *  2. If this is an anonymous class that implement a singleton enum case,
    *     pass $name and $ordinal parameters to the enum superclass. The class
    *     looks like this:
    *
    *       class $anon extends E(...) {
    *          ...
-   *          private def $ordinal = N
-   *          override def toString = S
-   *          ...
    *       }
    *
    *     After the transform it is expanded to
    *
-   *       class $anon extends E(..., N, S) {
-   *          ...
-   *          "removed $ordinal and toString"
-   *          ...
+   *       class $anon extends E(..., $name, _$ordinal) { // if class implements a simple enum case
+   *          "same as before"
+   *       }
+   *
+   *       class $anon extends E(..., "A", 0) { // if class implements a value enum case `A` with ordinal 0
+   *          "same as before"
    *       }
    */
   override def transformTemplate(templ: Template)(using Context): Template = {
     val cls = templ.symbol.owner
     if cls.derivesFromJavaEnum then
+      registerEnumClass(cls)
```

Comment:
This seem to rely on the enum class appearing before the enum values in the AST, this should be documented.

Reply:
I've added some comments and a test